### ISSUE TITLE
[FR] Make mlflow tracking backend scale beyond thousand of runs

### ISSUE THREAD

**Issue by rom1504** [mlflow/mlflow#2730]
## Describe the proposal 
Mlflow tracking backend currently is unable (with any storage backend) to scale in retrieval of runs beyond thousand of runs. My proposal is to consider this as an important issue for mlflow and to prioritize pull requests and issues that goes in this direction.
Here are some ideas : 
* being able to retrieve only some of the metrics, tags, params https://github.com/mlflow/mlflow/pull/2212 => improve the backend speed by about x15 but still slow for more than 20000 runs
* using a more efficient storage backend that is adapted to document storage (elastic search, graphql, ...) => that may be a real solution that could truly scale
* I would be glad to hear about other ideas

### Motivation
The use case would be supporting machine learning practitioners that run as teams or alone thousand of runs for a single experiment. Being able to retrieve these runs and compare them is quite important in this case.

### Proposed Changes
The changes would affect the storage backend and possibly the backend API.

The objective of this issue is to hear what you think about this ? is it something you consider important for databricks ? are there lot of users of mlflow that have this use case ?

**Comment by dbczumar** [comment]
@rom1504 can you elaborate on the specific APIs that exhibit excessive latency with large numbers of runs? E.g., search with a large number of max results, search with a small number of max results?

Regarding #2212, I think that @mparkhe's suggestion of adding a simpler, stateless `columns` parameter to the Search API seems like a perfectly reasonable mechanism for improving search performance (see https://github.com/mlflow/mlflow/pull/2212/files#discussion_r362662626). @mparkhe do you have opinions about the viability of this API? Is it something we'd be open to introducing sooner than later?

**Comment by mparkhe** [comment]
Hi @rom1504,

I am a bit confused about this part

> Mlflow tracking backend currently is unable (with any of the currently supported storage backend) to scale in retrieval of runs beyond thousand of runs

With pagination support you should be able to retrieve the data you need from the backend a few hundred-thousand runs at a time. The limitation would be on how to manage the storage in the calling client (if you intend to keep all the runs in memory) but I would not categorize of this as a scale issue with backend. Indeed, with file based backed 1000s of runs will be an issue but we have seen SQL based backend scale for such use cases.

### Projection of specific `columns`
I agree with @dbczumar 's comment above. It would make sense to add a columns field to `SearchRuns` proto. The proposed schema from [the RFC](https://docs.google.com/document/d/1Ygti9oyMbi4DJNbaM0D49nNIKb8BLXwGYQ7WmrOKJ3s/edit?usp=sharing) was something like this--

```
message SearchRuns {
  option (scalapb.message).extends = "com.databricks.rpc.RPC[$this.Response]";

  // List of experiment IDs to search over.
  repeated string experiment_ids = 1;

  // A filter expression over params, metrics, and tags, that allows returning a subset of
  // runs. The syntax is a subset of SQL that supports ANDing together binary operations
  // between a param, metric, or tag and a constant.
  //
  // Example: ``metrics.rmse < 1 and params.model_class = 'LogisticRegression'``
  //
  // You can select columns with special characters (hyphen, space, period, etc.) by using double quotes:
  // ``metrics."model class" = 'LinearRegression' and tags."user-name" = 'Tomas'``
  //
  // Supported operators are ``=``, ``!=``, ``>``, ``>=``, ``<``, and ``<=``.
  optional string filter = 4;

  // Whether to display only active, only deleted, or all runs.
  // Defaults to only active runs.
  optional ViewType run_view_type = 3 [default = ACTIVE_ONLY];

  // Maximum number of runs desired. Max threshold is 50000
  optional int32 max_results = 5 [default = 1000];

  // List of columns to be ordered by, including attributes, params, metrics, and tags with an
  // optional "DESC" or "ASC" annotation, where "ASC" is the default.
  // Example: ["params.input DESC", "metrics.alpha ASC", "metrics.rmse"]
  // Tiebreaks are done by start_time DESC followed by run_id for runs with the same start time
  // (and this is the default ordering criterion if order_by is not provided).
  repeated string order_by = 6;

  optional string page_token = 7;

  // Return only these metrics, params, and tags with the response.
  // If unspecified all info and data columns are returned.
  // Example: ["metrics.accuracy", "metrics.mse", "params.layers", "tags.project"]
  repeated string columns = 8;

  message Response {
    // Runs that match the search criteria.
    repeated Run runs = 1;
    optional string next_page_token = 2;
  }
}
```

However, this is a massive change and it would require writing up a design proposal on backend implementation on how this columns field will be used for projection of specific columns from the backend storage. 

Specifically answer some design tradeoffs around -

1. Parsing the `columns` proto field, specifically if there are any difference with the mechanisms used for parsing `order_by` field.
2. Case-sensitivity? Eg: if some runs specify a metric as "MSE" vs others as "mse"
3. How is the field translated to project columns using specific backends: FileStore, SQL store -- for the latter what mechanisms will be used to inject columns name the query such as to avoid SQL injection attack on backend servers
4. Query design and handling query performance and tradeoffs if a users specify 1000s of column names with the `SearchRuns` API call.
5. Testing plan and validation tests
6. Performance tradeoffs: some additional complexity of projecting specific columns vs faster runtimes when streaming back fewer bytes

**Comment by rom1504** [comment]
@dbczumar 
> @rom1504 can you elaborate on the specific APIs that exhibit excessive latency with large numbers of runs? E.g., search with a large number of max results, search with a small number of max results?

Yes, I'm talking about retrieving a lot of runs (say 30000 (this is not a theoretical case, we actually have that)). For this as far as I know the only way is search_runs.
The use case for retrieving that many runs is mainly correlation analysis (checking which parameter gets good result in some metrics, checking that 2 metrics are correlated, ...)

> With pagination support you should be able to retrieve the data you need from the backend a few hundred-thousand runs at a time. 

We don't observe that. In our case when using the sql storage (30000 runs with each having hundred of metrics, tags, params) it takes more than 15min (when increasing the timeout) to retrieve the runs.


For the whitelisting PRs, if the main issue is to 1 list of columns instead of 3 and writing some doc/design doc on how this works, that's definitely something that we can be done. I think we could consider doing that.

**Comment by dbczumar** [comment]
Hi @rom1504, it would be great to profile the slow query operations that you're experiencing. Perhaps there are addressable bottlenecks in the execution of SQL queries for search. Would you mind performing some profiling and providing feedback? https://docs.sqlalchemy.org/en/13/faq/performance.html outlines approaches for profiling SqlAlchemy operations.

**Comment by smurching** [comment]
@rom1504 thanks for filing, just wanted to check in - were you able to profile the SQL queries / do you have a reproducible example to facilitate further investigation?
